### PR TITLE
Implements a seamless login flow that remembers your last sign in scope

### LINKED ISSUE #172: Improve Login Flow
Good enough:
  - [x] Change the copy of the "Sign in" button to an icon button with the octocat and "Sign in with GitHub" - @discorick 
  - [x] /login Simplify the two step process with just [Sign in with GitHub] - @discorick 
  - [x] /login Add (Green icon) Your email to the permissions asked for - @discorick 
  - [x] "Sign in with github button" Change it's default behavior to "public" scope + email - @discorick 
  - [x] Force a re-auth for all _new_ "login" clicks
  - [ ] Move public vs private to separate OAuth applications - @rauhryan
  - [ ] Store credentials in couchdb (must be encrypted)
  - [ ] Change copy on pricing page to say we store tokens with XYZ encryption and never look at your data without your expressed permission
  - [x] Ensure that email that is collected is making it into intercom [#1 criteria for acceptance] 
  - [x] Sense public auth only on the dashboard and provide an upgrade option - @discorick 

Better:
  - [ ] Skip /login all together and default to "Email scope" 
  - [ ] [DASHBOARD]  If we only have the "email" scope, the landing page should be a "Welcome" here's how you get started similar to what [/login] does today. 
  - [x] [DASHBOARD] If we know they only have Public access, have an option to upgrade to private ("Not seeing your private repositories? [Upgrade access] [?]"

### PR DESCRIPTION
**Pull request by rauhryan**
Closes https://github.com/huboard/huboard-web/issues/168
Closes #172 
Closes https://github.com/huboard/huboard/issues/624

### CODE REVIEW

File: config/environments/production.rb [hunk 1]
```diff
@@ -76,6 +76,8 @@
   # Use default logging formatter so that PID and timestamp are not suppressed.
   config.log_formatter = ::Logger::Formatter.new
 
+  config.middleware.use Rack::Attack
```
Comment (dahlbyk) [inline config/environments/production.rb]: Why move?

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
   def current_user
-    github_user(:private) || github_user(:default) || OpenStruct.new
+    github_user(:default) || github_user(:private) || github_user(:public) || OpenStruct.new
   end
   def controller? *controller
     (controller.include?(params[:controller]) || controller.include?(params[:action])) ? "nav__btn--active nav__item--current": ''
   end
   def user_token
-    current_user ? current_user.token : nil
+    current_user.token
```
Comment (dahlbyk) [inline app/helpers/application_helper.rb]: Why drop the guard?

File: config/environments/production.rb (hunk 1 above)
Comment (rauhryan) [inline config/environments/production.rb]: I cleans up the logs for development. 

Don't really need the log noise while I'm developing. 

Sorry I threw that in there, just felt compeled

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (rauhryan) [inline app/helpers/application_helper.rb]: Don't need the guard, it was unreachable code. 

the current_user fallback to the an empty OpenStruct that will return nil for token

**Comment by dahlbyk** [comment]
Probably a trivial fix that's worth just including here, but https://huboard-rails-pr-291.herokuapp.com/dashboard throws a 500 if you're not authenticated. Should challenge for auth.

**Comment by dahlbyk** [comment]
Is there anything we can do about this?
chrome_2016-05-05_16-50-40

Repro:
1. `/login/private`
2. `/login/public`
3. Note public-only warning, even though private repo is visible.

**Comment by dahlbyk** [comment]
Also, thoughts on directing the Private tab to `/login/private` instead of disabling it?

**Comment by dahlbyk** [comment]
A few other questions, just thinking through edge cases:

1. What's the value of keeping both `/login` and `/welcome`? Should `/login` redirect to `/login/github`?
2. Should `/welcome` redirect to `/dashboard` if a non-default scope is available?

Really just trying to think through how we could completely shut down access routes to weird states.

**Comment by discorick** [comment]
> Also, thoughts on directing the Private tab to /login/private instead of disabling it?

I played with it originally, but it is a deceiving UX. I.E:

- User is in public
- User wants to see private too so they click private tab
- GitHub upgrades Oauth to private
- User is redirected to private repo only filter... wonders what happened to their public repos

**Comment by dahlbyk** [comment]
> I played with it originally, but it is a deceiving UX.

:+1:

**Comment by rauhryan** [comment]
@dahlbyk new push addresses your edge cases. 

I think we are ready to 🚢

**Comment by discorick** [comment]
Just one minor UX nitpick if you think its worth it, we could probably move everything closer to the top of the browser on welcome:

image

**Comment by rauhryan** [comment]
> Just one minor UX nitpick if you think its worth it, we could probably move everything closer to the top of the browser on welcome:

Nitpick noted. 

I'm going to leave it for now, I'm good to ship the way it is

Anything else before we :ship:?

**Comment by discorick** [comment]
It may be worth implementing @dahlbyk 's suggestion that if there is a private or publics scope already, `/welcome` should redirect to `/dashboard` 

This would effectively block users from reaching bad states from scope downgrades.

**Comment by dahlbyk** [comment]
I can still reproduce https://github.com/huboard/huboard-web/pull/291#issuecomment-218192754 if I have authorized private and hit `/login/public` (via `/login`).

Also noticed that if the app is revoked, `/dashboard` redirects to `/login`.

**Comment by rauhryan** [comment]
> Also noticed that if the app is revoked, /dashboard redirects to /login.

This is because Ghee::Unauthorized will log you out and redirect to `/login`, I can easily change this behavior... ref: https://github.com/huboard/huboard-web/blob/master/app/controllers/application_controller.rb#L13-L19

> I can still reproduce #291 (comment) if I have authorized private and hit /login/public (via /login).

We can still ship, knowing that "could" happen. 

Currently there is only one scenario where the application redirects or links to `/login`

I'm not ready to completely kill `/login`, I'd like to keep it around in case we need to bring it back. 

I think the question is: 

  1. What's the risk or do we care if users "url hack" the `/login` route?
  2. Are we ok to ship, knowing there are edge cases that can get you into a weird state? 

If someone gets into a funky state, the solution is... 

/logout => /login/github

The user will be taken to either /welcome or /dashboard where there are links to upgrade access that will fix it. 

Without adding datastore or switch auth frameworks, I don't think there is a perfect solution. 

My vote is: 

 Make a decision on how we handle `Ghee::Unauthorized` and ship it

**Comment by discorick** [comment]
> Make a decision on how we handle `Ghee::Unauthorized`

Just to send users to `/login/github` ?

**Comment by dahlbyk** [comment]
> I'm not ready to completely kill `/login`, I'd like to keep it around in case we need to bring it back.

That's why we have Git. :wink:

But seriously, life gets simpler if we make `/login` redirect to `/login/github`. We can easily remove the redirect if `/login` needs to come back.


With that plus a `/welcome ` to `/dashboard` redirect I think we are more than sufficiently buttoned up. I just don't want to have to revisit this again for a long time.

**Comment by rauhryan** [comment]
OK! I've added the requested changes. 

Let's :ship: this bad boy!

**Comment by dahlbyk** [comment]
👏 :sparkles: :clap: Excellent work!

**Pull request merged**